### PR TITLE
fix: Changed MultiHeadAttention's implementation from a naive one to pytorch's native one

### LINKED ISSUE #873: MultiHeadAttention scaling factor
Hello,

Forgive me if this is intentional, in `braindecode/modules/attention.py` line 754 for the `MultiHeadAttention` class, the scaling factor is

`scaling = self.emb_size ** (1 / 2)`

when it should probably be

`scaling = (self.emb_size // self.num_heads) ** (1 / 2)`, as per the multihead case.

(This class  is currently used for the ctnet and eegconformer models).

Furthermore, this class might benefit from wrapping PyTorch's native implementation to support features like FlashAttention.

### PR DESCRIPTION
**Pull request by Leob000**
Hello,
This is intended to fix #873 

Two "breaking changes" different from the initial naive implementation:
- The scaling factor becomes `(self.emb_size // self.num_heads) ** (1 / 2)`
- The mask is inverted (the initial implementation used a mask where True meant to keep the value while Pytorch uses True to ignore the value). Braincode doesn't seem to use this mask in other parts of the codebase to I made a direct change without guardrails, I'm not sure how to handle that otherwise.

I made a quick standalone file to compare the difference in output of the two methods (with a fixed scaling factor for the naive method). Caveat: as the weight initialization differ, a manual copy of the initial weights is made between the two methods.

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from einops.layers.torch import Rearrange


class MultiHeadAttention_new(nn.Module):
    def __init__(self, emb_size, num_heads, dropout=0.0):
        super().__init__()
        self.mha = nn.MultiheadAttention(
            emb_size, num_heads, dropout=dropout, batch_first=True
        )

    def forward(self, x, mask=None):
        return self.mha(x, x, x, attn_mask=mask, need_weights=False)[0]


class MultiHeadAttention_old(nn.Module):
    def __init__(self, emb_size, num_heads, dropout=0):
        super().__init__()
        self.emb_size = emb_size
        self.num_heads = num_heads
        self.keys = nn.Linear(emb_size, emb_size)
        self.queries = nn.Linear(emb_size, emb_size)
        self.values = nn.Linear(emb_size, emb_size)
        self.att_drop = nn.Dropout(dropout)
        self.projection = nn.Linear(emb_size, emb_size)

        self.rearrange_stack = Rearrange(
            "b n (h d) -> b h n d",
            h=num_heads,
        )
        self.rearrange_unstack = Rearrange(
            "b h n d -> b n (h d)",
        )

    def forward(self, x, mask=None):
        queries = self.rearrange_stack(self.queries(x))
        keys = self.rearrange_stack(self.keys(x))
        values = self.rearrange_stack(self.values(x))
        energy = torch.einsum("bhqd, bhkd -> bhqk", queries, keys)
        if mask is not None:
            fill_value = float("-inf")
            energy = energy.masked_fill(~mask, fill_value)

        # scaling = self.emb_size ** (1 / 2)
        scaling = (self.emb_size // self.num_heads) ** (1 / 2)
        att = F.softmax(energy / scaling, dim=-1)
        att = self.att_drop(att)
        out = torch.einsum("bhal, bhlv -> bhav ", att, values)
        out = self.rearrange_unstack(out)
        out = self.projection(out)
        return out


def run_comparison():
    print("--- Comparing Old vs New Attention ---")
    B, L, E, H = 4, 10, 16, 4
    torch.manual_seed(42)

    mha_old = MultiHeadAttention_old(E, H)
    mha_new = MultiHeadAttention_new(E, H)

    # Sync weights
    with torch.no_grad():
        w_q, b_q = mha_old.queries.weight, mha_old.queries.bias
        w_k, b_k = mha_old.keys.weight, mha_old.keys.bias
        w_v, b_v = mha_old.values.weight, mha_old.values.bias
        mha_new.mha.in_proj_weight.copy_(torch.cat([w_q, w_k, w_v], dim=0))
        mha_new.mha.in_proj_bias.copy_(torch.cat([b_q, b_k, b_v], dim=0))
        mha_new.mha.out_proj.weight.copy_(mha_old.projection.weight)
        mha_new.mha.out_proj.bias.copy_(mha_old.projection.bias)

    x = torch.randn(B, L, E)
    x_old = x.clone().detach().requires_grad_(True)
    x_new = x.clone().detach().requires_grad_(True)

    # Test 1: Forward (No Mask)
    out_old = mha_old(x_old)
    out_new = mha_new(x_new)
    diff = (out_old - out_new).abs()
    print(
        f"Forward Diff (No Mask): Max: {diff.max().item():.2e} | Mean: {diff.mean().item():.2e}"
    )

    # Test 2: Backward (No Mask)
    target = torch.randn_like(out_old)
    F.mse_loss(out_old, target).backward()
    F.mse_loss(out_new, target).backward()
    assert x_old.grad is not None and x_new.grad is not None
    diff_grad = (x_old.grad - x_new.grad).abs()
    print(
        f"Grad Diff (No Mask):    Max: {diff_grad.max().item():.2e} | Mean: {diff_grad.mean().item():.2e}"
    )

    # Test 3: Forward (With Mask - Raw / Wrong Convention)
    mask = torch.ones(L, L, dtype=torch.bool)
    mask[:, -3:] = False  # Mask last 3 positions
    out_old_masked = mha_old(x_old, mask=mask)

    # This demonstrates that Old uses True=Keep while New uses True=Ignore
    out_new_raw_mask = mha_new(x_new, mask=mask)
    diff_raw = (out_old_masked - out_new_raw_mask).abs()
    print(
        f"Forward Diff (Masked, Raw):      Max: {diff_raw.max().item():.2e} | Mean: {diff_raw.mean().item():.2e} (High diff confirms mask convention mismatch)"
    )

    # Test 4: Forward (With Mask - Inverted / Correct Convention)
    # Old: True=Keep, New: True=Ignore (so we pass ~mask)
    out_new_masked = mha_new(x_new, mask=~mask)
    diff_masked = (out_old_masked - out_new_masked).abs()
    print(
        f"Forward Diff (Masked, Inverted): Max: {diff_masked.max().item():.2e} | Mean: {diff_masked.mean().item():.2e}"
    )


if __name__ == "__main__":
    run_comparison()
```

Output
```
--- Comparing Old vs New Attention ---
Forward Diff (No Mask): Max: 8.94e-08 | Mean: 1.94e-08
Grad Diff (No Mask):    Max: 4.37e-10 | Mean: 7.25e-11
Forward Diff (Masked, Raw):      Max: 4.84e-01 | Mean: 1.63e-01 (High diff confirms mask convention mismatch)
Forward Diff (Masked, Inverted): Max: 8.94e-08 | Mean: 1.96e-08
```

### CODE REVIEW

**Comment by bruAristimunha** [comment]
Can you just run the benchmark example from the EEG Conformer PR?

**Comment by bruAristimunha** [comment]
https://github.com/braindecode/braindecode/pull/454

**Comment by Leob000** [comment]
Sure, here is the benchmark:
# Loss curves
- PR Implementation (native pytorch):
native_training_curves

- Original naive implementation with a fixed scaling (scaling = (self.emb_size // self.num_heads) ** (1 / 2)):
oldfixed_training_curves

- Original naive implementation with the original scaling (scaling = self.emb_size ** (1 / 2)):
oldnonfixed_training_curves


# Confusion matrices
- PR Implementation (native pytorch):
native_confusion_matrix

- Original naive implementation with a fixed scaling (scaling = (self.emb_size // self.num_heads) ** (1 / 2)):
oldfixed_confusion_matrix

- Original naive implementation with the original scaling (scaling = self.emb_size ** (1 / 2)):
oldnonfixed_confusion_matrix

# Code
I used the following code (with some tweaks compared to #454 else it wouldn't work), and changed the MultiHeadAttention class directly in `braindecode/modules/attention.py` to test the three different cases.

```python
import matplotlib.pyplot as plt
import pandas as pd
import torch
from matplotlib.lines import Line2D
from numpy import multiply
from sklearn.metrics import confusion_matrix
from skorch.callbacks import LRScheduler
from skorch.helper import predefined_split
from pathlib import Path

from braindecode import EEGClassifier
from braindecode.datasets import MOABBDataset
from braindecode.models import EEGConformer
from braindecode.preprocessing import (
    Preprocessor,
    create_windows_from_events,
    exponential_moving_standardize,
    preprocess,
)
from braindecode.util import set_random_seeds
from braindecode.visualization import plot_confusion_matrix

subject_id = 3
dataset = MOABBDataset(dataset_name="BNCI2014_001", subject_ids=[subject_id])


low_cut_hz = 4.0  # low cut frequency for filtering
high_cut_hz = 38.0  # high cut frequency for filtering
# Parameters for exponential moving standardization
factor_new = 1e-3
init_block_size = 1000
# Factor to convert from V to uV
factor = 1e6

preprocessors = [
    Preprocessor("pick_types", eeg=True, meg=False, stim=False),  # Keep EEG sensors
    Preprocessor(lambda data: multiply(data, factor)),  # Convert from V to uV
    Preprocessor("filter", l_freq=low_cut_hz, h_freq=high_cut_hz),  # Bandpass filter
    Preprocessor(
        exponential_moving_standardize,  # Exponential moving standardization
        factor_new=factor_new,
        init_block_size=init_block_size,
    ),
]

# Transform the data
preprocess(dataset, preprocessors)


trial_start_offset_seconds = -0.5
# Extract sampling frequency, check that they are same in all datasets
sfreq = dataset.datasets[0].raw.info["sfreq"]
assert all([ds.raw.info["sfreq"] == sfreq for ds in dataset.datasets])
# Calculate the trial start offset in samples.
trial_start_offset_samples = int(trial_start_offset_seconds * sfreq)

# Create windows using braindecode function for this. It needs parameters to define how
# trials should be used.
windows_dataset = create_windows_from_events(
    dataset,
    trial_start_offset_samples=trial_start_offset_samples,
    trial_stop_offset_samples=0,
    preload=True,
)


splitted = windows_dataset.split("session")
train_set = splitted["0train"]
valid_set = splitted["1test"]


cuda = torch.cuda.is_available()  # check if GPU is available, if True chooses to use it
mps = torch.backends.mps.is_available()
device = "cuda" if cuda else "mps" if mps else "cpu"
if cuda:
    torch.backends.cudnn.benchmark = True

seed = 20200220
set_random_seeds(seed=seed, cuda=cuda)

n_classes = 4
# Extract number of chans and time steps from dataset
n_chans = train_set[0][0].shape[0]
input_window_samples = windows_dataset[0][0].shape[1]

model = EEGConformer(
    n_chans=n_chans,
    n_outputs=n_classes,
    final_fc_length=2760,
)

# Send model to GPU
if cuda:
    model.cuda()
elif mps:
    model.to("mps")


# These values we found good for shallow network:
lr = 0.0625 * 0.01
weight_decay = 0

# For deep4 they should be:
# lr = 1 * 0.01
# weight_decay = 0.5 * 0.001

batch_size = 64
n_epochs = 100

clf = EEGClassifier(
    model,
    criterion=torch.nn.CrossEntropyLoss,
    optimizer=torch.optim.AdamW,
    train_split=predefined_split(valid_set),  # using valid_set for validation
    optimizer__lr=lr,
    optimizer__weight_decay=weight_decay,
    batch_size=batch_size,
    max_epochs=n_epochs,
    classes=[0, 1, 2, 3],
    callbacks=[
        "accuracy",
        ("lr_scheduler", LRScheduler("CosineAnnealingLR", T_max=n_epochs - 1)),
    ],
    device=device,
)
# Model training for a specified number of epochs. `y` is None as it is already supplied
# in the dataset.
clf.fit(train_set, y=None)


# Extract loss and accuracy values for plotting from history object
results_columns = ["train_loss", "valid_loss", "train_accuracy", "valid_accuracy"]
df = pd.DataFrame(
    clf.history[:, results_columns],
    columns=results_columns,  # type:ignore
    index=clf.history[:, "epoch"],
)

# get percent of misclass for better visual comparison to loss
df = df.assign(
    train_misclass=100 - 100 * df.train_accuracy,
    valid_misclass=100 - 100 * df.valid_accuracy,
)

# plt.style.use("seaborn")
output_dir = Path("temp")
output_dir.mkdir(parents=True, exist_ok=True)
fig, ax1 = plt.subplots(figsize=(8, 3))
df.loc[:, ["train_loss", "valid_loss"]].plot(
    ax=ax1, style=["-", ":"], marker="o", color="tab:blue", legend=False, fontsize=14
)

ax1.tick_params(axis="y", labelcolor="tab:blue", labelsize=14)
ax1.set_ylabel("Loss", color="tab:blue", fontsize=14)

ax2 = ax1.twinx()  # instantiate a second axes that shares the same x-axis

df.loc[:, ["train_misclass", "valid_misclass"]].plot(
    ax=ax2, style=["-", ":"], marker="o", color="tab:red", legend=False
)
ax2.tick_params(axis="y", labelcolor="tab:red", labelsize=14)
ax2.set_ylabel("Misclassification Rate [%]", color="tab:red", fontsize=14)
ax2.set_ylim(ax2.get_ylim()[0], 85)  # make some room for legend
ax1.set_xlabel("Epoch", fontsize=14)
plt.title("Model")
# where some data has already been plotted to ax
handles = []
handles.append(
    Line2D([0], [0], color="black", linewidth=1, linestyle="-", label="Train")
)
handles.append(
    Line2D([0], [0], color="black", linewidth=1, linestyle=":", label="Valid")
)
plt.legend(handles, [h.get_label() for h in handles], fontsize=14)
plt.tight_layout()
fig.savefig(output_dir / "training_curves.png", dpi=150)
plt.close(fig)


# generate confusion matrices
# get the targets
y_true = valid_set.get_metadata().target
y_pred = clf.predict(valid_set)

# generating confusion matrix
confusion_mat = confusion_matrix(y_true, y_pred)

# add class labels
if hasattr(windows_dataset.datasets[0], "windows"):
    label_dict = windows_dataset.datasets[0].windows.event_id.items()
    labels = list(dict(sorted(list(label_dict), key=lambda kv: kv[1])).keys())
else:
    labels = sorted(list(set(windows_dataset.datasets[0].raw.annotations.description)))

# plot the basic conf. matrix
plot_confusion_matrix(confusion_mat, class_names=labels)
confusion_fig = plt.gcf()
confusion_fig.savefig(output_dir / "confusion_matrix.png", dpi=150)
plt.close(confusion_fig)
```

**Comment by PierreGtch** [comment]
Hi @Leob000, Thanks for your contribution and this nice analysis 🚀

Could you plot the three implementations in the same figure, as different hue? And classification error and loss as different figure to keep things less cluttered 

Thank you!

**Comment by Leob000** [comment]
I remade the plots, to get a better idea of the uncertainty I did multiple runs: for each method, 3 different subjects of the dataset, each having 3 runs with a different seed. (The code for the plots is on [this branch](https://github.com/Leob000/braindecode/tree/native-multiheadattention-benchmark) for reference).

Strangely the native torch way seem to behave a bit worse on the Validation.

loss_comparison

misclass_comparison

**Comment by bruAristimunha** [comment]
image

**Comment by bruAristimunha** [comment]
Hi @Leob000,

I've been studying attention more, and it seems the key here is the initialization. The way it's defined leads to better results. I tested it with about 3 Moabb datasets, and nothing conclusive.

I tried studying and benchmarking to use the layer, but everything affected the behavior or changed something. In the end, the best trade-off I found was to use what Torch uses to compute attention.

That way, we get the best of both worlds :)

**Comment by bruAristimunha** [comment]
many thanks for push what @agramfort suggested, we didn't replace for native layer, but we are much more faster now :)
